lib/util/stream_utils.js

```diff
@@ -30,55 +30,117 @@ shaka.util.StreamUtils = class {
    * @param {number} preferredAudioChannelCount
    */
   static chooseCodecsAndFilterManifest(manifest, preferredAudioChannelCount) {
-    const MimeUtils = shaka.util.MimeUtils;
+    const StreamUtils = shaka.util.StreamUtils;
 
     // To start, consider a subset of variants based on audio channel
     // preferences.
     // For some content (#1013), surround-sound variants will use a different
     // codec than stereo variants, so it is important to choose codecs **after**
     // considering the audio channel config.
-    const variants = shaka.util.StreamUtils.filterVariantsByAudioChannelCount(
+    const variants = StreamUtils.filterVariantsByAudioChannelCount(
         manifest.variants, preferredAudioChannelCount);
 
-    /**
-     * @param {!shaka.extern.Variant} variant
-     * @return {string}
-     */
-    function variantCodecs(variant) {
-      // Only consider the base of the codec string.  For example, these should
-      // both be considered the same codec: avc1.42c01e, avc1.4d401f
-      let baseVideoCodec = '';
-      if (variant.video) {
-        baseVideoCodec = MimeUtils.getCodecBase(variant.video.codecs);
-      }
+    // Now organize variants into buckets by codecs.
+    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+    let variantsByCodecs = StreamUtils.getVariantsByCodecs(variants);
+    variantsByCodecs = StreamUtils.filterVariantsByHeight(variantsByCodecs);
+
+    const bestCodecs = StreamUtils.findBestCodecs(variantsByCodecs);
 
-      let baseAudioCodec = '';
-      if (variant.audio) {
-        baseAudioCodec = MimeUtils.getCodecBase(variant.audio.codecs);
+    // Filter out any variants that don't match, forcing AbrManager to choose
+    // from the most efficient variants possible.
+    manifest.variants = manifest.variants.filter((variant) => {
+      const codecs = StreamUtils.getGroupVariantCodecs(variant);
+      if (codecs == bestCodecs) {
+        return true;
       }
 
-      return baseVideoCodec + '-' + baseAudioCodec;
-    }
+      shaka.log.debug('Dropping Variant (better codec available)', variant);
+      return false;
+    });
+  }
 
-    // Now organize variants into buckets by codecs.
-    /** @type {!shaka.util.MultiMap.<shaka.extern.Variant>} */
+  /**
+  * Get variants by codecs.
+  *
+  * @param {!Array<shaka.extern.Variant>} variants
+  * @return {!shaka.util.MultiMap.<shaka.extern.Variant>}
+  */
+  static getVariantsByCodecs(variants) {
```

Comment:
This is only used internally by this class, so it should end with an underscore and be marked with `@private`.

By the way, I really like the refactor you've done to split this into multiple methods.  It will really help with readability.